### PR TITLE
Fixed missing callbacks for pending events

### PR DESCRIPTION
**Pull request by e-hndrks**
When setting a listener, any pending (i.e. unhandled) event received before would not result in an immediate callback. Now the set_listener call also checks for pending events, and invokes the registered callbacks when appropriate. This fix is a prerequisite for #https://github.com/eclipse-cyclonedds/cyclonedds-cxx/issues/410

### CODE REVIEW

**Review by eboasson — COMMENTED** [review]
I think this'll work pretty well for setting the status callbacks on the reader/writer, but I don't think it'll cause listener callbacks when setting a listeners on a entity higher up in the hierarchy. Extending it to deal with the data-available/data-on-readers events seems quite doable to me, and so we should add that.

There are also some other details and it really needs some (stress) testing ... Testing this will be tricky, there are two aspects to it:
* whether it covers all listeners
* whether the invocation mechanism works properly
* whether it indeed handles cases where listeners are (re)set to 0 correctly

The first one can realistically be covered by inspection (just check whether each status flag/listener is accounted for and the names line up).

The second is probably most easily covered by testing with data-available and data-on-readers (those are easy to manipulate via the API, after all).

The last one could also be done with just one type of event and then checking all cases are handled consistently.

Especially that second test will take quite a bit of effort. Perhaps it would be an idea to add a sanity-check test in this PR and do a follow-up that does more thorough checking later? That sanity-check should be straightforward.

File: src/core/ddsc/src/dds__types.h
Comment (e-hndrks) [reply src/core/ddsc/src/dds__types.h:259]: Check. Typical consequence of using copy-paste too eagerly. Will correct.

File: src/core/ddsc/src/dds__reader.h
Comment (e-hndrks) [reply src/core/ddsc/src/dds__reader.h:26]: Again a matter of copy-pasting too eagerly. Will correct this.

File: src/core/ddsc/src/dds_entity.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_entity.c:1064]: Good point. I will make sure that the listeners affected by `pushdown_listener` will also be evaluating the status to see if they need to invoke the registered callback.

File: src/core/ddsc/src/dds_entity.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_entity.c:1073]: Ah, that makes sense: I already wondered in which case the `reset_on_invoke` would not be set when the corresponding listener callback was. So instead I will be checking whether a callback function is registered prior to determining whether it needs to be invoked or not.

File: src/core/ddsc/src/dds_entity.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_entity.c:1029]: Makes sense. Which documentation files would be impacted by this?

File: src/core/ddsc/src/dds_entity.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_entity.c:1071]: In my 2nd iteration that might not be needed anymore, because the evaluation of the status and the callback still happen with the `m_cb_count` set. So other listener calls should not be able to interfere.

File: src/core/ddsc/src/dds_reader.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_reader.c:448]: Now added a check to see if a callback has been registered prior to checking the status.

**Comment by trittsv** [comment]
Hey @eboasson and @e-hndrks :) is there a chance that after the merge of this PR (and the cxx counterpart) a new CycloneDDS release could be created?

We are in a pitty because we want to release our software which is using cyclonedds. But our software does not work without these fixes.
We just want to say the users that they should install official CycloneDDS Version X.Y. And not our patched CycloneDDS version.

**Comment by eboasson** [comment]
@trittsv definitely!

The details are (of course) still to be decided ...

My wished-for 11.0 release is long overdue, in part because of the paperwork to be done and the need for a release/progress review by the Eclipse Foundation. The former is not a valid argument, but the latter really does mean it takes a while longer, and so altogether probably a longer timescale than what would work for you.

I also had a quick look into backporting this to 0.10.x and that looks eminently feasible. There are some conflicts but they appear straightforward. The behavioural change is small enough, and more importantly, entirely in line with how people would've worked around it (always invoking the listener function once in case an event happened earlier) and leaving that workaround in should cause no problems.

So let's aim for that.

File: src/core/ddsc/src/dds_entity.c
Comment (eboasson) [reply src/core/ddsc/src/dds_entity.c:1035]: Only `dds_set_listener` in `dds.h`. It just needs a bit of text like:

> `dds_set_listener` will invoke any listeners for which the corresponding status flag is set. It may cause spurious invocations, including multiple invocations for one listener. For most cases this is unlikely, but for the `DATA_ON_READERS` listeners it is quite likely, though not certain.

At least that would work for me, but YMMV. Suggestions always welcome 🙂

File: src/core/ddsc/src/dds_entity.c
Comment (eboasson) [inline src/core/ddsc/src/dds_entity.c:1018]: Would it not be better to also increment/decrement `m_cb_count`? I can't (yet) give an execution why it could fail without it, but at the very least there is a consistency and the fact that shaving off a pair of trivial instructions in such a rarely used piece of code isn't worth an inconsistency.

File: src/core/ddsc/src/dds_entity.c
Comment (eboasson) [inline src/core/ddsc/src/dds_entity.c:1013]: This line (or, perhaps more precisely, the fact that it is held here until after the `dds_entity_deriver_invoke_cbs_for_pending_events` means in this one path it holds the parent entity's `m_observers_lock` while invoking the listener. I'm sure that'll be fine in many cases, but before I can merge it I want to understand what the effect of that will be.

Or it needs to be released before invoking the listeners. The only thing it is doing for which it requires it is the `dds_override_inherited_listener` call. So moving the "unlock" is probably the best option.

File: src/core/ddsc/tests/deadline.c
Comment (eboasson) [inline src/core/ddsc/tests/deadline.c:30]: It is pretty trivial to split this commit into one that fixes the deadline test and one that makes the functional changes on setting the listeners. I think doing so would be worth it.

**Review by eboasson — COMMENTED** [review]
Just a few details, but overall it looks really good — thanks! I've marked some conversations as "resolved" as an easy way to remove some clutter and to signal my agreement.

File: src/core/ddsc/src/dds__reader.h
Comment (e-hndrks) [reply src/core/ddsc/src/dds__reader.h:25]: Will copy this into my pull request.

File: src/core/ddsc/src/dds__types.h
Comment (e-hndrks) [reply src/core/ddsc/src/dds__types.h:122]: Will copy this into my pull request.

File: src/core/ddsc/src/dds_entity.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_entity.c:1015]: What you are basically saying here is that since the events only occur on the ultimate children, it doesn't hurt to let for example a Participant or a Publisher check its OfferedIncompatibleQosStatus bit, since that bit will never be set on that level? That might be true, and the check might actually prevent a pending DATA_ON_READERS to be invoked, because a subscriber probably does have children..... 
I will remove this check in my pull-request.

File: src/core/ddsc/src/dds_entity.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_entity.c:1018]: In the revised pull-request we will increase m_cb_pending_count instead, which gives increased safety against two threads running dds_set_listener on the same entity in parallel.

File: src/core/ddsc/src/dds_entity.c
Comment (e-hndrks) [reply src/core/ddsc/src/dds_entity.c:1013]: In the next pull-request we will unlock m_observers_lock prior to invoking the `dds_entity_deriver_invoke_cbs_for_pending_events`.

File: src/core/ddsc/tests/deadline.c
Comment (e-hndrks) [reply src/core/ddsc/tests/deadline.c:30]: I guess that means we will do the deadline fix in a separate pull request?

**Review by eboasson — APPROVED** [review]
It looks good to me now 😀 Thank you very much @e-hndrks!

**Pull request merged**
